File: design-to-spec/scripts/validate-output.py
#!/usr/bin/env python3
"""Validate generated design-to-spec output files against the YAML contracts."""

from __future__ import annotations

import argparse
import json
import re
import subprocess
import sys
from pathlib import Path
from typing import Any

try:
    import yaml
except ImportError:
    yaml = None


EVENT_PATTERN = re.compile(r"[a-z]+(?:-[a-z]+)+")
HEADING_PATTERN = re.compile(r"^(#{1,6})\s+(.+?)\s*$", re.MULTILINE)


def load_yaml(path: Path) -> dict[str, Any]:
    if yaml is not None:
        try:
            data = yaml.safe_load(path.read_text(encoding="utf-8"))
        except yaml.YAMLError as exc:
            raise ValueError(f"{path}: invalid YAML: {exc}") from exc
    else:
        data = load_yaml_with_ruby(path)
    if not isinstance(data, dict):
        raise ValueError(f"{path}: top-level YAML must be a mapping")
    return data


def load_yaml_with_ruby(path: Path) -> Any:
    script = "require 'yaml'; require 'json'; puts JSON.generate(YAML.load_file(ARGV[0]))"
    try:
        result = subprocess.run(
            ["ruby", "-e", script, str(path)],
            check=True,
            text=True,
            capture_output=True,
        )
    except FileNotFoundError as exc:
        raise ValueError(f"{path}: PyYAML is unavailable and Ruby fallback is not installed") from exc
    except subprocess.CalledProcessError as exc:
        detail = exc.stderr.strip() or exc.stdout.strip()
        raise ValueError(f"{path}: invalid YAML: {detail}") from exc
    return json.loads(result.stdout)


def read_text(path: Path) -> str:
    if not path.exists():
        raise ValueError(f"{path}: file does not exist")
    return path.read_text(encoding="utf-8")


def collect_endpoint_urls(api_doc: dict[str, Any]) -> dict[str, str]:
    endpoints = api_doc.get("api", {}).get("endpoints", []) or []
    return {
        endpoint.get("id"): endpoint.get("url")
        for endpoint in endpoints
        if endpoint.get("id") and endpoint.get("url")
    }


def collect_event_names(mapping_doc: dict[str, Any]) -> set[str]:
    events: set[str] = set()
    for binding in mapping_doc.get("mapping", {}).get("bindings", []) or []:
        if binding.get("direction") != "ui_to_event":
            continue
        target_event = binding.get("target_event", "")
        events.update(EVENT_PATTERN.findall(target_event))
        transform = binding.get("transform", "")
        events.update(EVENT_PATTERN.findall(transform))
    return events


def collect_open_questions(api_doc: dict[str, Any], mapping_doc: dict[str, Any]) -> list[dict[str, str]]:
    questions: list[dict[str, str]] = []
    questions.extend(api_doc.get("api", {}).get("open_questions", []) or [])
    questions.extend(mapping_doc.get("mapping", {}).get("open_questions", []) or [])
    return questions


def normalized_probe(text: str, length: int = 12) -> str:
    compact = re.sub(r"\s+", "", text)
    compact = re.sub(r"[`\"'()（）？?，,。.:：；;、]", "", compact)
    return compact[:length]


def section_text(markdown: str, heading: str, level: int = 2) -> str:
    target = "#" * level
    matches = list(HEADING_PATTERN.finditer(markdown))
    for index, match in enumerate(matches):
        if match.group(1) == target and match.group(2).strip() == heading:
            start = match.end()
            end = len(markdown)
            for next_match in matches[index + 1 :]:
                if len(next_match.group(1)) <= level:
                    end = next_match.start()
                    break
            return markdown[start:end].strip()
    return ""


def extract_numbered_items(markdown_section: str) -> dict[int, str]:
    items: dict[int, str] = {}
    current_number: int | None = None
    current_lines: list[str] = []

    def flush() -> None:
        if current_number is not None:
            items[current_number] = " ".join(line.strip() for line in current_lines).strip()

    for line in markdown_section.splitlines():
        match = re.match(r"^\s*(\d+)\.\s+(.*)$", line)
        if match:
            flush()
            current_number = int(match.group(1))
            current_lines = [match.group(2)]
        elif current_number is not None and line.startswith((" ", "\t")):
            current_lines.append(line)
    flush()
    return items


def extract_table_rows(markdown_section: str) -> list[list[str]]:
    rows: list[list[str]] = []
    for line in markdown_section.splitlines():
        stripped = line.strip()
        if not stripped.startswith("|") or not stripped.endswith("|"):
            continue
        cells = [cell.strip() for cell in stripped.strip("|").split("|")]
        if cells and all(re.fullmatch(r":?-{3,}:?", cell) for cell in cells):
            continue
        rows.append(cells)
    return rows


def collect_state_ids_from_notes(notes_text: str) -> set[str]:
    rows = extract_table_rows(section_text(notes_text, "状态枚举"))
    state_ids: set[str] = set()
    for row in rows[1:]:
        if not row:
            continue
        match = re.search(r"`([^`]+)`", row[0])
        if match:
            state_ids.add(match.group(1))
    return state_ids


def collect_tracking_anchor_ids(notes_text: str) -> set[str]:
    rows = extract_table_rows(section_text(notes_text, "埋点锚点"))
    anchors: set[str] = set()
    for row in rows[1:]:
        for cell in row:
            anchors.update(EVENT_PATTERN.findall(cell))
    return anchors


def collect_trace_ids(markdown: str) -> set[str]:
    return set(re.findall(r"`((?:component|binding|state|request):[^`]+)`", markdown))


def validate_traceability(
    ui_doc: dict[str, Any],
    mapping_doc: dict[str, Any],
    notes_text: str,
    data_fetching_text: str,
    spec_text: str,
) -> list[str]:
    errors: list[str] = []
    trace_section = section_text(notes_text, "Traceability")
    if not trace_section:
        return errors

    notes_trace_ids = collect_trace_ids(trace_section)
    data_fetching_trace_ids = collect_trace_ids(data_fetching_text)
    spec_trace_ids = collect_trace_ids(spec_text)

    for component in ui_doc.get("ui", {}).get("components", []) or []:
        component_id = component.get("id")
        if component_id and f"component:{component_id}" not in notes_trace_ids:
            errors.append(f"notes.md Traceability is missing component trace component:{component_id}")

    for index, binding in enumerate(mapping_doc.get("mapping", {}).get("bindings", []) or [], start=1):
        direction = binding.get("direction")
        trace_id = f"binding:{index}:{direction}"
        if direction and trace_id not in notes_trace_ids:
            errors.append(f"notes.md Traceability is missing binding trace {trace_id}")

    for state in ui_doc.get("ui", {}).get("states", []) or []:
        state_id = state.get("id")
        trace_id = f"state:{state_id}"
        if state_id and trace_id not in notes_trace_ids:
            errors.append(f"notes.md Traceability is missing state trace {trace_id}")
        if state.get("required") is True and state_id and trace_id not in spec_trace_ids:
            errors.append(f"spec.md is missing required state trace {trace_id}")

    for request in mapping_doc.get("mapping", {}).get("data_fetching", {}).get("requests", []) or []:
        request_id = request.get("id")
        trace_id = f"request:{request_id}"
        if request_id and trace_id not in data_fetching_trace_ids:
            errors.append(f"data-fetching.md is missing request trace {trace_id}")

    return errors


def question_number(question: dict[str, str]) -> int | None:
    question_id = str(question.get("id", ""))
    match = re.search(r"(\d+)$", question_id)
    return int(match.group(1)) if match else None


def question_is_anchored(question: dict[str, str], numbered_questions: dict[int, str], notes_text: str) -> bool:
    number = question_number(question)
    if number is not None and number in numbered_questions:
        return True

    content = question.get("content", "")
    probe = normalized_probe(content)
    normalized_notes = normalized_probe(notes_text, length=len(notes_text))
    if probe and probe in normalized_notes:
        return True

    # Fallback: require several meaningful tokens to appear somewhere in the
    # open-question section. This tolerates copy edits while still catching
    # genuinely dropped questions.
    open_question_text = "\n".join(numbered_questions.values())
    tokens = [
        token
        for token in re.split(r"[`\s\"'()（）？?，,。.:：；;、/]+", content)
        if len(token) >= 4
    ]
    if not tokens:
        return False
    hits = sum(1 for token in tokens if token in open_question_text)
    return hits >= min(2, len(tokens))


def validate_outputs(
    ui_doc: dict[str, Any],
    api_doc: dict[str, Any],
    mapping_doc: dict[str, Any],
    notes_text: str,
    data_fetching_text: str,
    spec_text: str,
) -> tuple[list[str], list[str]]:
    errors: list[str] = []
    warnings: list[str] = []

    required_fragments = ["## ADDED Requirements", "### Requirement:", "#### Scenario:", "- WHEN", "- THEN"]
    if "## MODIFIED Requirements" in spec_text:
        required_fragments[0] = "## MODIFIED Requirements"
    for fragment in required_fragments:
        if fragment not in spec_text:
            errors.append(f"spec.md is missing OpenSpec fragment {fragment!r}")

    notes_state_ids = collect_state_ids_from_notes(notes_text)
    for state in ui_doc.get("ui", {}).get("states", []) or []:
        if state.get("required") is True:
            state_id = state.get("id", "")
            if state_id and state_id not in notes_state_ids:
                errors.append(f"required state {state_id!r} is not listed in notes.md 状态枚举")
            if state_id and state_id not in spec_text:
                errors.append(f"required state {state_id!r} is not mentioned in spec.md")
            assertion = state.get("render_assertion", "")
            if assertion and assertion not in spec_text and state_id not in spec_text:
                warnings.append(f"required state {state_id!r} render_assertion is not reflected in spec.md")

    endpoint_urls = collect_endpoint_urls(api_doc)
    for request in mapping_doc.get("mapping", {}).get("data_fetching", {}).get("requests", []) or []:
        endpoint_id = request.get("endpoint")
        endpoint_url = endpoint_urls.get(endpoint_id)
        if endpoint_id and endpoint_url and endpoint_url not in data_fetching_text:
            errors.append(f"data-fetching.md does not mention endpoint {endpoint_url!r} for request {request.get('id')!r}")

    tracking_anchor_ids = collect_tracking_anchor_ids(notes_text)
    combined_notes_spec = notes_text + "\n" + spec_text
    for event_name in collect_event_names(mapping_doc):
        if event_name not in tracking_anchor_ids and event_name not in combined_notes_spec:
            errors.append(f"event {event_name!r} is not mentioned in notes.md or spec.md")

    numbered_questions = extract_numbered_items(section_text(notes_text, "开放问题"))
    for question in collect_open_questions(api_doc, mapping_doc):
        priority = question.get("priority", "")
        if priority == "P0" and not question_is_anchored(question, numbered_questions, notes_text):
            warnings.append(f"P0 open question may be missing from notes.md: {question.get('content', '')}")

    if "## 待确认项汇总" not in data_fetching_text:
        warnings.append("data-fetching.md is missing a 待确认项汇总 section")
    if "## 开放问题" not in notes_text:
        errors.append("notes.md is missing an 开放问题 section")
    if "## 埋点锚点" not in notes_text:
        warnings.append("notes.md is missing an 埋点锚点 section")

    errors.extend(validate_traceability(ui_doc, mapping_doc, notes_text, data_fetching_text, spec_text))

    return errors, warnings


def main() -> int:
    parser = argparse.ArgumentParser(description="Validate generated design-to-spec output files.")
    parser.add_argument("--ui", required=True, type=Path, help="Path to ui-schema.yaml")
    parser.add_argument("--api", required=True, type=Path, help="Path to api-schema.yaml")
    parser.add_argument("--mapping", required=True, type=Path, help="Path to mapping-logic.yaml")
    parser.add_argument("--notes", required=True, type=Path, help="Path to notes.md")
    parser.add_argument("--data-fetching", required=True, type=Path, help="Path to data-fetching.md")
    parser.add_argument("--spec", required=True, type=Path, help="Path to spec.md")
    parser.add_argument("--strict", action="store_true", help="Treat warnings as errors")
    args = parser.parse_args()

    try:
        ui_doc = load_yaml(args.ui)
        api_doc = load_yaml(args.api)
        mapping_doc = load_yaml(args.mapping)
        notes_text = read_text(args.notes)
        data_fetching_text = read_text(args.data_fetching)
        spec_text = read_text(args.spec)
        errors, warnings = validate_outputs(ui_doc, api_doc, mapping_doc, notes_text, data_fetching_text, spec_text)
    except ValueError as exc:
        print(str(exc), file=sys.stderr)
        return 1

    for warning in warnings:
        print(f"WARN: {warning}", file=sys.stderr)
    if args.strict:
        errors.extend(warnings)

    if errors:
        for error in errors:
            print(f"ERROR: {error}", file=sys.stderr)
        return 1

    print("OK: output files are valid")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: design-to-spec/scripts/validate-output.js
#!/usr/bin/env node
// Validate generated design-to-spec output files against the YAML contracts.

import { existsSync, readFileSync } from "node:fs";
import { fileURLToPath } from "node:url";
import { loadYaml } from "./lib/yaml.js";
import { parseArgs, requireOpts } from "./lib/cli.js";

const EVENT_PATTERN = /[a-z]+(?:-[a-z]+)+/g;
const HEADING_PATTERN = /^(#{1,6})\s+(.+?)\s*$/gm;

function pyRepr(value) {
  if (typeof value === "string") return `'${value}'`;
  if (value === null || value === undefined) return "None";
  return String(value);
}

function readText(path) {
  if (!existsSync(path)) {
    throw new Error(`${path}: file does not exist`);
  }
  return readFileSync(path, "utf8");
}

function collectEndpointUrls(apiDoc) {
  const map = new Map();
  for (const endpoint of apiDoc.api?.endpoints ?? []) {
    if (endpoint.id && endpoint.url) map.set(endpoint.id, endpoint.url);
  }
  return map;
}

function collectEventNames(mappingDoc) {
  const events = new Set();
  for (const binding of mappingDoc.mapping?.bindings ?? []) {
    if (binding.direction !== "ui_to_event") continue;
    for (const text of [binding.target_event ?? "", binding.transform ?? ""]) {
      const matches = String(text).match(EVENT_PATTERN) ?? [];
      for (const m of matches) events.add(m);
    }
  }
  return events;
}

function collectOpenQuestions(apiDoc, mappingDoc) {
  return [
    ...(apiDoc.api?.open_questions ?? []),
    ...(mappingDoc.mapping?.open_questions ?? []),
  ];
}

function normalizedProbe(text, length = 12) {
  let compact = text.replace(/\s+/g, "");
  compact = compact.replace(/[`"'()（）？?，,。.:：；;、]/g, "");
  return compact.slice(0, length);
}

function findAllHeadings(markdown) {
  // global regex with capture - need to use matchAll
  const re = /^(#{1,6})\s+(.+?)\s*$/gm;
  const matches = [];
  let m;
  while ((m = re.exec(markdown)) !== null) {
    matches.push({ start: m.index, end: m.index + m[0].length, level: m[1].length, title: m[2].trim() });
  }
  return matches;
}

function sectionText(markdown, heading, level = 2) {
  const matches = findAllHeadings(markdown);
  for (let i = 0; i < matches.length; i++) {
    const m = matches[i];
    if (m.level === level && m.title === heading) {
      const start = m.end;
      let end = markdown.length;
      for (let j = i + 1; j < matches.length; j++) {
        if (matches[j].level <= level) {
          end = matches[j].start;
          break;
        }
      }
      return markdown.slice(start, end).trim();
    }
  }
  return "";
}

function extractNumberedItems(markdownSection) {
  const items = new Map();
  let currentNumber = null;
  let currentLines = [];

  function flush() {
    if (currentNumber !== null) {
      const joined = currentLines.map((l) => l.trim()).join(" ").trim();
      items.set(currentNumber, joined);
    }
  }

  for (const line of markdownSection.split(/\r?\n/)) {
    const m = /^\s*(\d+)\.\s+(.*)$/.exec(line);
    if (m) {
      flush();
      currentNumber = parseInt(m[1], 10);
      currentLines = [m[2]];
    } else if (currentNumber !== null && (line.startsWith(" ") || line.startsWith("\t"))) {
      currentLines.push(line);
    }
  }
  flush();
  return items;
}

function extractTableRows(markdownSection) {
  const rows = [];
  for (const line of markdownSection.split(/\r?\n/)) {
    const stripped = line.trim();
    if (!stripped.startsWith("|") || !stripped.endsWith("|")) continue;
    const cells = stripped.replace(/^\|/, "").replace(/\|$/, "").split("|").map((c) => c.trim());
    if (cells.length > 0 && cells.every((c) => /^:?-{3,}:?$/.test(c))) continue;
    rows.push(cells);
  }
  return rows;
}

function collectStateIdsFromNotes(notesText) {
  const rows = extractTableRows(sectionText(notesText, "状态枚举"));
  const stateIds = new Set();
  for (let i = 1; i < rows.length; i++) {
    const row = rows[i];
    if (!row || row.length === 0) continue;
    const m = /`([^`]+)`/.exec(row[0]);
    if (m) stateIds.add(m[1]);
  }
  return stateIds;
}

function collectTrackingAnchorIds(notesText) {
  const rows = extractTableRows(sectionText(notesText, "埋点锚点"));
  const anchors = new Set();
  for (let i = 1; i < rows.length; i++) {
    for (const cell of rows[i]) {
      const matches = cell.match(EVENT_PATTERN) ?? [];
      for (const m of matches) anchors.add(m);
    }
  }
  return anchors;
}

function collectTraceIds(markdown) {
  const result = new Set();
  const re = /`((?:component|binding|state|request):[^`]+)`/g;
  let m;
  while ((m = re.exec(markdown)) !== null) result.add(m[1]);
  return result;
}

function validateTraceability(uiDoc, mappingDoc, notesText, dataFetchingText, specText) {
  const errors = [];
  const traceSection = sectionText(notesText, "Traceability");
  if (!traceSection) return errors;

  const notesTraceIds = collectTraceIds(traceSection);
  const dataFetchingTraceIds = collectTraceIds(dataFetchingText);
  const specTraceIds = collectTraceIds(specText);

  for (const component of uiDoc.ui?.components ?? []) {
    const id = component.id;
    if (id && !notesTraceIds.has(`component:${id}`)) {
      errors.push(`notes.md Traceability is missing component trace component:${id}`);
    }
  }

  (mappingDoc.mapping?.bindings ?? []).forEach((binding, idx) => {
    const direction = binding.direction;
    const traceId = `binding:${idx + 1}:${direction}`;
    if (direction && !notesTraceIds.has(traceId)) {
      errors.push(`notes.md Traceability is missing binding trace ${traceId}`);
    }
  });

  for (const state of uiDoc.ui?.states ?? []) {
    const id = state.id;
    const traceId = `state:${id}`;
    if (id && !notesTraceIds.has(traceId)) {
      errors.push(`notes.md Traceability is missing state trace ${traceId}`);
    }
    if (state.required === true && id && !specTraceIds.has(traceId)) {
      errors.push(`spec.md is missing required state trace ${traceId}`);
    }
  }

  for (const request of mappingDoc.mapping?.data_fetching?.requests ?? []) {
    const id = request.id;
    const traceId = `request:${id}`;
    if (id && !dataFetchingTraceIds.has(traceId)) {
      errors.push(`data-fetching.md is missing request trace ${traceId}`);
    }
  }

  return errors;
}

function questionNumber(question) {
  const id = String(question.id ?? "");
  const m = /(\d+)$/.exec(id);
  return m ? parseInt(m[1], 10) : null;
}

function questionIsAnchored(question, numberedQuestions, notesText) {
  const number = questionNumber(question);
  if (number !== null && numberedQuestions.has(number)) return true;

  const content = question.content ?? "";
  const probe = normalizedProbe(content);
  const normalizedNotes = normalizedProbe(notesText, notesText.length);
  if (probe && normalizedNotes.includes(probe)) return true;

  const openQuestionText = [...numberedQuestions.values()].join("\n");
  const tokens = content
    .split(/[`\s"'()（）？?，,。.:：；;、/]+/)
    .filter((t) => t.length >= 4);
  if (tokens.length === 0) return false;
  const hits = tokens.reduce((acc, t) => acc + (openQuestionText.includes(t) ? 1 : 0), 0);
  return hits >= Math.min(2, tokens.length);
}

function validateOutputs(uiDoc, apiDoc, mappingDoc, notesText, dataFetchingText, specText) {
  const errors = [];
  const warnings = [];

  let requiredFragments = ["## ADDED Requirements", "### Requirement:", "#### Scenario:", "- WHEN", "- THEN"];
  if (specText.includes("## MODIFIED Requirements")) {
    requiredFragments = ["## MODIFIED Requirements", ...requiredFragments.slice(1)];
  }
  for (const fragment of requiredFragments) {
    if (!specText.includes(fragment)) {
      errors.push(`spec.md is missing OpenSpec fragment ${pyRepr(fragment)}`);
    }
  }

  const notesStateIds = collectStateIdsFromNotes(notesText);
  for (const state of uiDoc.ui?.states ?? []) {
    if (state.required === true) {
      const id = state.id ?? "";
      if (id && !notesStateIds.has(id)) {
        errors.push(`required state ${pyRepr(id)} is not listed in notes.md 状态枚举`);
      }
      if (id && !specText.includes(id)) {
        errors.push(`required state ${pyRepr(id)} is not mentioned in spec.md`);
      }
      const assertion = state.render_assertion ?? "";
      if (assertion && !specText.includes(assertion) && !specText.includes(id)) {
        warnings.push(`required state ${pyRepr(id)} render_assertion is not reflected in spec.md`);
      }
    }
  }

  const endpointUrls = collectEndpointUrls(apiDoc);
  for (const request of mappingDoc.mapping?.data_fetching?.requests ?? []) {
    const endpointId = request.endpoint;
    const endpointUrl = endpointUrls.get(endpointId);
    if (endpointId && endpointUrl && !dataFetchingText.includes(endpointUrl)) {
      errors.push(
        `data-fetching.md does not mention endpoint ${pyRepr(endpointUrl)} for request ${pyRepr(request.id)}`,
      );
    }
  }

  const trackingAnchorIds = collectTrackingAnchorIds(notesText);
  const combinedNotesSpec = notesText + "\n" + specText;
  for (const eventName of collectEventNames(mappingDoc)) {
    if (!trackingAnchorIds.has(eventName) && !combinedNotesSpec.includes(eventName)) {
      errors.push(`event ${pyRepr(eventName)} is not mentioned in notes.md or spec.md`);
    }
  }

  const numberedQuestions = extractNumberedItems(sectionText(notesText, "开放问题"));
  for (const question of collectOpenQuestions(apiDoc, mappingDoc)) {
    const priority = question.priority ?? "";
    if (priority === "P0" && !questionIsAnchored(question, numberedQuestions, notesText)) {
      warnings.push(`P0 open question may be missing from notes.md: ${question.content ?? ""}`);
    }
  }

  if (!dataFetchingText.includes("## 待确认项汇总")) {
    warnings.push("data-fetching.md is missing a 待确认项汇总 section");
  }
  if (!notesText.includes("## 开放问题")) {
    errors.push("notes.md is missing an 开放问题 section");
  }
  if (!notesText.includes("## 埋点锚点")) {
    warnings.push("notes.md is missing an 埋点锚点 section");
  }

  errors.push(...validateTraceability(uiDoc, mappingDoc, notesText, dataFetchingText, specText));

  return { errors, warnings };
}

function main() {
  let args;
  try {
    args = parseArgs(process.argv.slice(2), {
      options: ["ui", "api", "mapping", "notes", "data-fetching", "spec"],
      flags: ["strict"],
    });
    requireOpts(args, ["ui", "api", "mapping", "notes", "data-fetching", "spec"]);
  } catch (err) {
    console.error(err.message);
    return 1;
  }

  let result;
  try {
    const uiDoc = loadYaml(args.ui);
    const apiDoc = loadYaml(args.api);
    const mappingDoc = loadYaml(args.mapping);
    const notesText = readText(args.notes);
    const dataFetchingText = readText(args["data-fetching"]);
    const specText = readText(args.spec);
    result = validateOutputs(uiDoc, apiDoc, mappingDoc, notesText, dataFetchingText, specText);
  } catch (err) {
    console.error(err.message);
    return 1;
  }

  for (const warning of result.warnings) {
    console.error(`WARN: ${warning}`);
  }
  let errors = result.errors;
  if (args.strict) {
    errors = errors.concat(result.warnings);
  }

  if (errors.length > 0) {
    for (const error of errors) {
      console.error(`ERROR: ${error}`);
    }
    return 1;
  }

  console.log("OK: output files are valid");
  return 0;
}

if (process.argv[1] === fileURLToPath(import.meta.url)) {
  process.exit(main());
}

export { validateOutputs, sectionText, extractTableRows, extractNumberedItems, collectTraceIds };
